Rework on index.html, addImageryLayerFromJSON, addElevationLayerFromJSON, GUIGlobe

vcoindet opened the pull request:
**In ApiGlogbe.js :** 
- addElevationLayerFromJSON()
  Allows to load a single Imagery Layer from a JSON file
- addImageryLayersFromJSONArray()
  Allows to load Imagery Layers from an array composed of JSON files
- addImageryLayerFromJSON()
  Allows to load a single Imagery Layer from a JSON file
- addElevationLayersFromJSONArray() 
  Allows to load Elevation Layers from an array composed of JSON files

**GuiTools** 
In examples/GUI/
New class to create a new menu using dat.gui

**JS and JSON layers**
In examples/layers

**In index.html**
Rework on the all file using the previous functions. This makes the index.html file more readable...

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -100,14 +101,48 @@ function preprocessLayer(layer, provider) {
+ */
+
+ApiGlobe.prototype.addImageryLayerFromJSON = function(url) {
+     return this.readJSON(url).then(function(result){
```
Why not use  `JSONDriver.read(url)` directly instead of adding a one-liner function?

gchoqueux replied:
ok

peppsac commented on src/Core/Commander/Providers/JSON_Provider.js:
```diff
@@ -0,0 +1,13 @@
+function JSON_Provider(ellipsoid) {
+    //Constructor
+
+}
```
This file is not needed

gchoqueux replied:
removed

peppsac commented on examples/GUI/GUIGlobe.js:
```diff
@@ -0,0 +1,69 @@
+ */
+
+
+function GUIGlobe() {
```
Why _Globe_ suffix? This is a helper class to build a GUI to display/manipulate layers, so what do you think of `LayerGUI` ?

vcoindet replied:
It's not only to display layers, it's also used to create section in the menu. See the "addGUI" function which is used to create the 'RealisticLighting' section.
This class might be used to do other things in the future, so I'm not sure that "layerGUI" is a good name...

peppsac replied:
I don't really see the point of `GUIGlobe`.
What are the benefits of introduce this class over using `dat.gui` directly?
From my point of view, there's no need to introduce a new class if you just a new a few simple helpers methods in a single place...

peppsac commented on examples/GUI/GUIGlobe.js:
```diff
@@ -0,0 +1,69 @@
+    //Constructor
+
+    this.gui = new dat.GUI( { autoPlace : false});
+    this.gui.domElement.id = 'menuDiv';
```
Don't hardcode the element id. Add it as a parameter to the constructor instead.

peppsac commented on examples/GUI/GUIGlobe.js:
```diff
@@ -0,0 +1,69 @@
+function GUIGlobe() {
+    //Constructor
+
+    this.gui = new dat.GUI( { autoPlace : false});
```
`dat` will be undefined.
Maybe it's time to try to integrate it as a npm dev-dependency for itowns?

vcoindet replied:
I will probably need some help on that point... I guess there is some work to do on webpack.config et package.json but I'm not very comfortable with the package.json...

gchoqueux replied:
@tbroyer Need help to use dat.gui npm dependency

peppsac commented:
Instead of moving layers declaration content to separate files to cleanup index.html, supporting standard `GetCapabilities` method for WMTS layers would be a huge benefit.

We could then use:

``` js
            itowns.viewer.addImageryLayer({
                protocol:   "wmts",
                id:         "Ortho",
                url:        "http://wxs.ign.fr/va5orxd0pgzvq3jxutqfuy0b/geoportail/wmts",
                name: "ORTHOIMAGERY.ORTHOPHOTOS",
                updateStrategy: {
                    type: 0, /* see LayerUpdateStrategy.js */
                    options: {}
                }
            });
```

Instead of having to copy-paste 10s of config lines.

vcoindet commented:
I totally agree, supporting standard GetCapabilities method for WMTS layers would be a huge benefit.
But I think we should do both ways (GetCapabilities and the method proposed in this PR). I think it could be usefull for users

peppsac commented:
Sure. But it's a good opportunity to implement `GetCapabilities` :)

vcoindet commented:
Yes I totally agree ;)

gchoqueux committed:
addImageryLayersFromJSONArray()
addImageryLayerFromJSON()
addElevationLayersFromJSONArray()
addElevationLayerFromJSON()

GUIGlobe : Api GUI for globe.
Adding layer JSON Files in examples

gchoqueux commented:
branch updated and rebased

tbroyer commented on autodeploy.sh:
```diff
@@ -34,16 +34,22 @@ git clone $REPO --single-branch --branch $TARGET_BRANCH out
 # Copy demo
 cp index.html out/
+# Copy layers
+cp -R examples/layers out/
```
Won't that create an `out/layers/` rather than `out/examples/layers/`? (as expected by the `git rm` at lines 37 and 52)

gchoqueux replied:
updated
question @tbroyer : how use dat.gui and npm?

tbroyer commented:
> question @tbroyer : how use dat.gui and npm?

There's `dat-gui` for that 3 year-old (sic!) v0.5, or `dat.gui` for the more recent 0.6.0 and 0.6.1.
Given that it's for examples only, it'd have to be a `dev` dependency, not a standard one (the iTowns library itself doesn't depend on it); but it'd be useless to depend on it from NPM unless we also process examples with Webpack (using multiple `entry` entries), which may not be worth it.

vcoindet committed:
autodeploy with examples

vcoindet commented:
branch updated and rebased

peppsac commented on examples/GUI/GuiTools.js:
```diff
@@ -0,0 +1,71 @@
+
+		var folder = this.elevationGui.addFolder(layer.id);
+		folder.add( { frozen: false }, 'frozen').onChange(function(value) {
+			this.api.setLayerFreeze(this, value);
```
```
GuiTools.js:38 Uncaught TypeError: Cannot read property 'setLayerFreeze' of undefined
```

vcoindet replied:
Sorry, I forgot to correct this one... Will do it

vcoindet replied:
done

peppsac commented on examples/layers/JSLayers/IGN_MNT.js:
```diff
@@ -0,0 +1,73 @@
+itowns.viewer.addElevationLayer({
```
AFAIK `JS/Layers/*.js` files aren't used. What's the point of adding them then?

vcoindet replied:
that's right, we should probably remove all the js layer files as we don't use it. I will do it !

vcoindet replied:
JS files removed. done

peppsac commented on examples/GUI/GuiTools.js:
```diff
@@ -0,0 +1,71 @@
+/* global viewerDiv*/
+
+function GuiTools(api,domId) {
+    //Constructor
```
Avoid useless comments. Same comment applies to the comments at the top of the file

vcoindet replied:
done

peppsac commented:
@vcoindet  didn't answer my question : 

> I don't really see the point of GUIGlobe.
> What are the benefits of introduce this class over using dat.gui directly?
> From my point of view, there's no need to introduce a new class if you just a new a few simple helpers methods in a single place...

vcoindet commented:
Well, the purpose here was to make a more clean index.html. I think the way we organise our menu/GUI should be just an exemple for users, I don't think we should impose our own menu/GUI.

vcoindet committed:
add addElevationLayerFromJSON()
add addElevationLayersFromJSONArray()
add addImageryLayerFromJSON()
add addImageryLayersFromJSONArray()

add GuiTools in examples/GUI/
add JSON layers in examples/layers/JSONLayers
Cleaning index.html

peppsac commented:
LGTM.
